src/oom/oomd-util.c

```diff
@@ -282,6 +283,204 @@ int oomd_cgroup_kill(const char *path, bool recurse, bool dry_run) {
+                const char *basename,
+                sd_json_variant *cparams,
+                OomdKillState *ks,
+                sd_event *e) {
+
+        _cleanup_(sd_varlink_close_unrefp) sd_varlink *link = NULL;
+        _cleanup_free_ char *hook_path = NULL;
+        int r;
+
+        assert(basename);
+        assert(cparams);
+        assert(e);
+        assert(ks);
+        assert(ks->manager);
+
+        log_info("Invoking oomd prekill hook %s for cgroup %s", basename, ks->ctx->path);
+
+        hook_path = path_join(VARLINK_ADDR_PATH_OOMD_PREKILL, basename);
+        if (!hook_path)
+                return log_oom_debug();
+
+        r = sd_varlink_connect_address(&link, hook_path);
+        if (r < 0)
+                return log_debug_errno(r, "Prekill socket at %s not connected, ignoring.", hook_path);
+
+        (void) sd_varlink_set_userdata(link, ks);
+        r = sd_varlink_set_description(link, "oomd prekill hook");
+        if (r < 0)
+                return log_debug_errno(r, "Failed to set varlink description: %m");
+        (void) sd_varlink_set_relative_timeout(link, ks->manager->prekill_timeout);
+
+        r = sd_varlink_attach_event(link, e, SD_EVENT_PRIORITY_NORMAL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to attach varlink to event loop: %m");
+
+        r = sd_varlink_bind_reply(link, prekill_callback);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to bind reply callback: %m");
+
+        r = sd_varlink_invoke(link, "io.systemd.oom.PreKill.Notify", cparams);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to call varlink method io.systemd.oom.PreKill.Notify: %m");
+
+        r = set_ensure_consume(&ks->links, NULL, TAKE_PTR(link));
+        if (r < 0)
+                return log_oom_debug();
+
+        oomd_kill_state_ref(ks);
+
+        return 0;
+}
+
+static int oomd_prekill_hook(Manager *m, OomdCGroupContext *ctx) {
+        _cleanup_closedir_ DIR *d = NULL;
+        int r;
+
+        assert(ctx);
+        assert(m);
+
+        if (m->prekill_timeout == 0) {
+                log_info("No oomd prekill timeout configured, skipping prekill hooks.");
+                return 0;
+        }
+
+        d = opendir(VARLINK_ADDR_PATH_OOMD_PREKILL);
+        if (!d) {
+                if (errno == ENOENT) {
+                        log_info("No prekill varlink socket directory %s, ignoring.", VARLINK_ADDR_PATH_OOMD_PREKILL);
+                        return 0;
+                }
+                return log_debug_errno(errno, "Failed to open prekill varlink socket directory %s: %m",
+                                         VARLINK_ADDR_PATH_OOMD_PREKILL);
+        }
+
+        _cleanup_(oom_kill_state_freep) OomdKillState *ks = new(OomdKillState, 1);
+        if (!ks)
+                return log_oom_debug();
+
+        *ks = (OomdKillState) {
+                .n_ref = 1,
+                .manager = m,
+                .ctx = ctx,
```

Comment:
shouldn't you refcount ctx here with `oomd_kill_state_ref()`? You're unrefing `ctx` in `oom_kill_state_free()`

Reply:
ah, I see it below at line 444. There is a bug though. If `!ks->ctx->path` or `sd_json_buildo()` would fail, it will cause a cleanup for `ks` => `oom_kill_state_freep()` => `oomd_cgroup_context_unref()` for not referenced ctx.